pennylane/math/fn.py

```diff
@@ -339,6 +492,39 @@ def get_interface(tensor):
     return TensorBox(tensor).interface
 
 
+def marginal_prob(prob, axis):
+    """Compute the marginal probability given joint probabilities of the full system.
+
+    Args:
+        prob (tensor_like): 1D tensor of probabilities. This tensor should of size
+            ``(2**N,)`` for some integer value ``N``.
+        axis (list[int]): the axis for which to calculate the marginal
+            probability distribution
+
+    Returns:
+        tensor_like: the marginal probabilities, of
+        size ``(2**len(axis),)``
+
+    **Example**
+
+    >>> x = tf.Variable([1, 0, 0, 1.], dtype=tf.float64) / np.sqrt(2)
+    >>> marginal_prob(x, axis=[0, 1])
+    <tf.Tensor: shape=(4,), dtype=float64, numpy=array([0.70710678, 0.        , 0.        , 0.70710678])>
+    >>> marginal_prob(x, axis=[0])
+    <tf.Tensor: shape=(2,), dtype=float64, numpy=array([0.70710678, 0.70710678])>
+    """
+    prob = flatten(prob)
+    num_wires = int(np.log2(len(prob)))
+
+    if num_wires == len(axis):
+        return prob
+
+    inactive_wires = tuple(set(range(num_wires)) - set(axis))
+    prob = reshape(prob, [2] * num_wires)

```

Comment:
Is this reshape to avoid copying? Or is something more going on?

Reply:
Oh, we are reshaping from a flat array of size 2**N (as returned from the QNode) to a N-dimensional array :) Makes it easier to simply sum over the axes we no longer care about.

But yes the reshape avoids copying as well.